text/0000-build-std/4-stage-1a.md

```diff
@@ -0,0 +1,1040 @@
+"never" ([?][rationale-build-std-never]) or "always", defaulting to "never":
+
+```toml
+[build]
+build-std = "always" # or `never`
+```
+
+`build-std` can also be specified in the `[target.<triple>]` and
+`[target.<cfg>]` sections ([?][rationale-build-std-target-section]):
+
+```toml
+[target.aarch64-unknown-illumos]
+build-std = "always" # or `never`
+```
+
+The `build-std` configuration locations have the following precedence
+([?][rationale-build-std-precedence]):
+
+1. `[target.<triple>]`
+2. `[target.<cfg>]`
+3. `[build]`
+
+As the Cargo configuration is local to the current installation of Cargo
+(typically in `~/.config/cargo`), the value of `build-std` is not influenced by
+the dependencies of the current crate.
+
+When `build-std` is set to "always", then the standard library will be
+unconditionally recompiled ([?][rationale-unconditional]) in the release profile
+defined in its workspace as part of every clean build
+([?][rationale-release-profile]). This is primarily useful for users of tier
+three targets. As with other dependencies, the standard library's build will
+respect the `RUSTFLAGS` environment variable.
+
+> [!NOTE]
+>
+> Configuration of the pre-built standard library is split across bootstrap and
+> the Cargo packages for the standard library. As much of this configuration as
+> possible should be moved to the Cargo profile for these packages so that the
+> artifacts produced by build-std match the pre-built standard library as much
+> as is feasible.
+
+Alongside `build-std`, a `build-std-crate` key will be introduced
+([?][rationale-build-std-crate]), which can be used to specify which crates from
+the standard library is to be built. Only "core", "alloc" and "std" are valid
+values for `build-std-crate`.
+
+```toml
+[build]
+build-std-crate = "std"
+```
+
+If [*Stage 1b* of this proposal][stage1b] is implemented then `build-std-crate`
+will not be used unless explicitly set and the crate graph's dependencies on the
+standard library will determine which crates are built instead. Otherwise,
+`build-std-crate` will default to "std".
+
+If `std` is to be built and Cargo is building a test using the default test
+harness then Cargo will also build the `test` crate.
+
+> [!NOTE]
+>
+> Inspired by the concept of [opaque dependencies][Opaque dependencies], the
+> standard library is resolved differently to other dependencies:
+>
+> - The lockfile included in the standard library source will be used when
+>   resolving the standard library's dependencies ([?][rationale-lockfile]).
+>
+> - The dependencies of the standard library crates are entirely opaque to the
+>   user. Different semver-compatible versions of these dependencies can
+>   exist in the user's resolve. The user cannot control compilation any of
+>   the dependencies of the `core`, `alloc` or `std` standard library crates
+>   individually (via profile overrides, for example).
+>
+> - The profile defined by the standard library will be used.
```

Comment:
Also one additional comment: I think that this will break LTO, because the stdlib will be compiled with `-Cembed-bitcode=no` (bootstrap explicitly overrides this to ship libstd that is LTO-compatible). But even without merging the stdlib and user's workspace profiles, this should be relatively to go around; cargo would just hardcode `-Cembed-bitcode=yes` if the user's profile has LTO enabled (or it would just use it always).